### ISSUE TITLE
Babel: investigate preprocessing static styles

### ISSUE THREAD

**Issue by necolas** [necolas/react-native-web#1132]
**Is your feature request related to a problem? Please describe.**

This feature request concerns further optimization of styles. A large number of styles used in components are static and will never be composed with styles from other components at runtime. This presents an optimization opportunity while preserving dynamic styling and style composition as a runtime feature. Along with moving the base View styles into the CSS "reset" (see #1136), this could further reduce the browser's scripting and layout timings on initial render.

Static style:

```js
<View style={styles.root} />

const styles = StyleSheet.create({
  height: 50
});
```

This is the most common use of styles and all the information required to resolve the styles is known at build time, without needing to know anything about the rest of the application.

**Describe a solution you'd like**

Perhaps the babel plugin can be extended to preprocess static styles. Since there are a [few style properties involved in bidirectional layout](https://github.com/necolas/react-native-web/blob/0.9.2/packages/react-native-web/src/exports/StyleSheet/i18nStyle.js#L52-L76) (e.g., `start`, `marginStart`, `transitionProperty`), and RTL layout can be toggled at runtime, these properties should be excluded. And it may not be worth processing styles that are used in a mixture of both composed and static context.

At build time, we can understand the difference between static styles not registered with `StyleSheet` (or inlined as object in render, tsk) and those where the value is determined at runtime. So, we can cheat and process the former with `StyleSheet` to resolve it to a `className`, which also saves us using the slowest resolve path at runtime (unregistered styles require the entire `style` object to be flattened and cannot rely on the cache for quick id-to-classNames transform).

For example:

```js
// dynamic styles (ignored)
// source
<View style={{ height: this.calculateHeight() }}} />
// compiled
<View style={{ height: this.calculateHeight() }}} />

// composed styles (ignored)
// source
<View style={[ styles.root1, this.props.style ]} />
// compiled
<View style={[ styles.root1, this.props.style ]} />

// static styles also used by composed (ignored??)
// source
<View style={styles.root1} />
// compiled
<View style={styles.root1} />

// static styles
// source
<View style={styles.root2} />
// compiled
<View className="rn-backgroundColor-2jss05 rn-height-im3hn" />

// static styles with bidi properties
// source
<View style={styles.root3} />
// compiled
<View className="rn-backgroundColor-2jss05 rn-height-im3hn" style={styles.root3} />

// "static" styles not registered with StyleSheet
// source
<View style={{ height: 50 }}} />
// compiled
<View className="rn-height-im3hn" />

// registered styles
// source
const styles = StyleSheet.create({
  root1: {
    backgroundColor: 'red',
    marginStart: 10,
    height: 50
  },
  root2: {
    backgroundColor: 'green',
    height: 50
  },
  root3, {
    backgroundColor: 'green',
    marginStart: 20,
    height: 50
  }
});
// compiled
const styles = StyleSheet.create({
  root1: {
    backgroundColor: 'red',
    marginStart: 10,
    height: 50
  },
  root3: {
    marginStart: 20
  }
});
```

It could be that an option is included to force processing of bidi properties, if you know you don't support RTL languages or layout at all.

The plugin would need to generate the CSS and inline it somewhere, probably with the rest of the predefined CSS in `StyleSheet`.

**Additional context**

A detail to be aware of is the integration with React DevTools. Currently, it's only enabled in development mode and since it relies on mapping style IDs to registered object, it would not expose the source styles that have been converted to classNames. Therefore, this work would be considered a production-only optimization.

Another idea to take this further: an option to extract the predefined CSS file from the client-side JS bundles entirely, and include it only on the server. This could remove a few KB of strings from the JS bundle if you knew that your server and service-worker would always provide the initial stylesheet inlined in the HTML shell document.

**Comment by giuseppeg** [comment]
hi @necolas I built an experimental library that can do something similar to what you are after. Please keep in mind that it is just an experiment and I built it for fun so it might be flawed. https://github.com/giuseppeg/style-sheet 

My library turns rules (object literals like `styles.root`) to arrays of atomic classes which have the following structure:

```css
dss_propertyName-value
```

At runtime I merge those arrays right to left using the `dss_propertyName` part of them (similarly to how Object.assign works).

This structure allows to extract static styles to file which is what the [babel plugin](https://github.com/giuseppeg/style-sheet/blob/master/src/babel.js) does. The plugin can also resolve static expressions and imported constants.

**Comment by giuseppeg** [comment]
Is it really bad to pre rendering the extracted styles via `link` or `style` tag? It seems that Twitter Lite's stylesheet is ~13kb after all. Anyway I believe that with a webpack loader (or dirty Babel side effects) you can move the shared files to common modules.

**Comment by necolas** [comment]
Open to looking at prototypes of this that reuse the existing client-side functions

**Comment by necolas** [comment]
Read the docs please

**Comment by giuseppeg** [comment]
@necolas not sure when I will have some time to make a prototype but my idea was to rewrite the module like this:

```js
import { rn1471scf } from './pathToCwd/.rn-tmp/preprocessed.css.js'
import { insertRuleOnce } from './sheet'

export default ({ children }) => {
   insertRuleOnce(rn1471scf) // can find other ways to insert the rule
   return <View style={{ __generated: 'rn-1471scf' }}>{children}</View>
}
```

Where `./pathToCwd/.rn-tmp/preprocessed.css.js` is generated at build time similarly to how linaria does https://github.com/callstack/linaria/blob/c949f19d76cb5027f980c3de864dae0b6f98ea2e/src/loader.js

webpack should pick up `./pathToCwd/.rn-tmp/preprocessed.css.js` and take care of the rest (bundling).

TBH I haven't thought about RTL and other dynamic values, not sure if you wanted to preprocess those (eg. margins)

**Comment by natew** [comment]
@necolas I've implemented static extraction just like this, with support even for conditionals and combined styles, but in order to work it needs an escape hatch to somehow pass in className. I've tried thinking through doing it with data-attributes, but it just adds incredible complexity.

For example, I'm just re-using react-native-web's internal atomic className generation, the same thing StyleSheet.create() generates. So to move over to data-attributes, we'd have to maintain some layer that then re-translates those to data attributes. And instead of having a single className prop, now have to insert/remove many properties.

React is great at escape hatches. I do think the `style={{ __generated: '' }}` pattern seems fine enough, though not generic at all. My preference would be something like:

```tsx
<View __domProps={{ className: '' }} />
```

I open this knowing you're not a fan of allowing any sort of escape hatch. But having found this issue, it seems to provide some "prior art". Plus, I have a working extractor that is easily adapted to "plain old StyleSheet.create" usage as well, it would be pretty nice to release it and get the community on board to improve.
